src/charm.py

```diff
@@ -1927,19 +1925,20 @@ def clean_ca_file_from_workload(self, secret_name: str) -> bool:
             logger.exception("CA file failed to clean. Error in config update")
             return False
 
-    def _reboot_on_detached_storage(self, event: EventBase) -> None:
-        """Reboot on detached storage.
+    def _check_detached_storage(self) -> None:
+        """Wait for storage to become available.
 
         Workaround for lxd containers not getting storage attached on startups.
 
         Args:
             event: the event that triggered this handler
         """
-        event.defer()
-        logger.error("Data directory not attached. Reboot unit.")
-        self.unit.status = WaitingStatus("Data directory not attached")
-        with contextlib.suppress(subprocess.CalledProcessError):
-            subprocess.check_call(["/usr/bin/systemctl", "reboot"])
+        for attempt in Retrying(stop=stop_after_attempt(10), wait=wait_fixed(1), reraise=True):
+            with attempt:
+                if not self._is_storage_attached():
+                    logger.error("Data directory not attached.")
+                    self.unit.status = WaitingStatus("Data directory not attached")
+                    raise StorageUnavailableError()
```

Comment:
Optional: Store the original unit status, as we do in the extensions code, and set it back if the storage becomes available before 10 seconds. Maybe it will speed up the unit status update.